### PR TITLE
[Feature] npu support torchair backend

### PR DESCRIPTION
**Pull request by XDaoHong**
## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

File: python/sglang/srt/model_executor/npu_graph_runner.py [hunk 1]
```diff
@@ -45,14 +55,14 @@
...
 
-@contextmanager
-def patch_model_npu(
-    model: torch.nn.Module,
-    enable_compile: bool,
-    num_tokens: int,
-    tp_group: GroupCoordinator,
-):
-    yield model
+# @contextmanager
```
Comment (eshoguli) [inline python/sglang/srt/model_executor/npu_graph_runner.py]: please, remove all comments

File: python/sglang/srt/model_executor/npu_graph_runner.py [hunk 2]
```diff
@@ -61,16 +71,73 @@ class NPUGraphRunner(CudaGraphRunner):
     def __init__(self, model_runner: ModelRunner):
         if model_runner.server_args.enable_torch_compile:
             patch_dynamo_context()
-        sglang.srt.model_executor.cuda_graph_runner.patch_model = patch_model_npu
+        # sglang.srt.model_executor.cuda_graph_runner.patch_model = patch_model_npu
         model_runner.attn_backend.enable_torch_compile = (
             model_runner.server_args.enable_torch_compile
         )
 
+        self.backend = os.getenv("NPU_TORCH_COMPILE_BACKEND", "reduce-overhead")
```
Comment (eshoguli) [inline python/sglang/srt/model_executor/npu_graph_runner.py]: rename `self.backend` to `self.backend_name`, please

File: python/sglang/srt/model_executor/npu_graph_runner.py [hunk 3]
```diff
@@ -61,16 +71,73 @@ class NPUGraphRunner(CudaGraphRunner):
...
+
         super().__init__(model_runner)
         self.use_mla = model_runner.model_config.attention_arch == AttentionArch.MLA
 
     def _create_device_graph(self):
         return torch.npu.NPUGraph()
+    
+    def capture(self) -> None:
```
Comment (eshoguli) [inline python/sglang/srt/model_executor/npu_graph_runner.py:76]: In my understanding: you don't need to implement capturing. Is it possible to use implemented in `CUDAGraphRunner` approach?

File: python/sglang/srt/model_executor/npu_graph_runner.py [hunk 4]
```diff
@@ -190,3 +273,149 @@ def replay(
         else:
             assert isinstance(output, PPProxyTensors)
             return PPProxyTensors({k: v[: self.bs] for k, v in output.tensors.items()})
+
+    def prepare_forward_batch(self, bs: int, num_tokens: int) -> ForwardBatch:
```
Comment (eshoguli) [inline python/sglang/srt/model_executor/npu_graph_runner.py:266]: Do you really need to prepare `ForwardBatch` instance yourself? Why you can not use implementation in `CUDAGraphRunner`?

File: python/sglang/srt/model_executor/npu_graph_runner.py (hunk 1 above)
Comment (XDaoHong) [reply python/sglang/srt/model_executor/npu_graph_runner.py]: done

File: python/sglang/srt/model_executor/npu_graph_runner.py (hunk 2 above)
Comment (XDaoHong) [reply python/sglang/srt/model_executor/npu_graph_runner.py]: done

File: python/sglang/srt/model_executor/npu_graph_runner.py (hunk 3 above)
Comment (XDaoHong) [reply python/sglang/srt/model_executor/npu_graph_runner.py:76]: NPUGraphRunner is inherited from CUDAGraphRunner. Capture is triggered during class initialization.
The backend of torchair integrates the capture and replay capabilities of npugraph. Therefore, torch.npu.graph does not need to be used to capture the compiled model.
To reduce modification, the self.capture method is rewritten to distinguish between torchair and npugraph logic.

File: python/sglang/srt/model_executor/npu_graph_runner.py (hunk 4 above)
Comment (XDaoHong) [reply python/sglang/srt/model_executor/npu_graph_runner.py:266]: prepare_forward_batch is used to prevent dynamo recompilation in the decode phase. In the torchair scenario, the decode phase invokes the compiled model instead of the graph obtained by using the compiled model capture. Therefore, ensure that the input in the warmup phase is the same as that in the decode phase.
In the warmup phase, the input of bs=2 is obtained from the tensor view whose bs=32. In the decode phase, bs=2 is not obtained from the view. Therefore, recompile is triggered.

File: python/sglang/srt/model_executor/npu_graph_runner.py (hunk 3 above)
Comment (eshoguli) [reply python/sglang/srt/model_executor/npu_graph_runner.py:76]: `NPUGraphRunner` (`npu_graph_runner.py`) & `CudaGraphRunner` (`cuda_graph_runner.py`) forward model with `torch.npu.NPUGraph` and `torch.cuda.CUDAGraph`. If torch AIR compilation backend doesn't require `NPUGraph` then you need another runner. For example: TorchAirRunner or something else. Let's discuss offline.

File: python/sglang/srt/model_executor/npu_graph_runner.py [hunk 5]
```diff
@@ -45,32 +55,78 @@
...
+    def capture(self) -> None:
+        # Reverse the order to enable better memory sharing across cuda graphs.
+        capture_range = (
+            tqdm.tqdm(list(reversed(self.capture_bs)))
+            if get_tensor_model_parallel_rank() == 0
+            else reversed(self.capture_bs)
+        )
+
+        graph_capture_context = empty_context()
+        if self.enable_torch_compile:
```
Comment (eshoguli) [inline python/sglang/srt/model_executor/npu_graph_runner.py:85]: Looks like your changes broke previous backend.
